x-pack/elastic-agent/pkg/core/server/server.go

```diff
@@ -0,0 +1,854 @@
+	firstRespChan := make(chan *proto.ActionResponse)
+	go func() {
+		// go func will not be leaked, because when the main function
+		// returns it will close the connection. that will cause this
+		// function to return.
+		observed, err := server.Recv()
+		if err != nil {
+			close(firstRespChan)
+			return
+		}
+		firstRespChan <- observed
+	}()
+
+	var ok bool
+	var firstResp *proto.ActionResponse
+	select {
+	case firstResp, ok = <-firstRespChan:
+		break
+	case <-time.After(InitialCheckinTimeout):
+		// close connection
+		s.logger.Debug("actions stream never sent initial response message; closing connection")
+		return nil
+	}
+	if !ok {
+		// close connection
+		return nil
+	}
+	if firstResp.Id != client.ActionResponseInitID {
+		// close connection
+		s.logger.Debug("actions stream first response message must be an init message; closing connection")
+		return status.Error(codes.InvalidArgument, "initial response must be an init message")
+	}
+	appState, ok := s.getByToken(firstResp.Token)
+	if !ok {
+		// no application with token; close connection
+		s.logger.Debug("actions stream sent an invalid token; closing connection")
+		return status.Error(codes.PermissionDenied, "invalid token")
+	}
+	appState.actionsLock.Lock()
+	if appState.actionsDone != nil {
+		// application is already connected (cannot have multiple); close connection
+		appState.actionsLock.Unlock()
+		s.logger.Debug("actions stream already exists for application; closing connection")
+		return status.Error(codes.AlreadyExists, "application already connected")
+	}
+	if !appState.actionsConn {
+		// application is being destroyed cannot reconnect; close connection
+		appState.actionsLock.Unlock()
+		s.logger.Debug("actions stream cannot connect, application is being destroyed; closing connection")
+		return status.Error(codes.Unavailable, "application cannot connect being destroyed")
+	}
+	actionsDone := make(chan bool)
+	appState.actionsDone = actionsDone
+	appState.actionsLock.Unlock()
+
+	defer func() {
+		appState.actionsLock.Lock()
+		appState.actionsDone = nil
+		appState.actionsLock.Unlock()
+	}()
+
+	// send the pending actions that need to be performed
+	recvDone := make(chan bool)
+	sendDone := make(chan bool)
+	go func() {
+		defer func() { close(sendDone) }()
+		for {
+			var pending *pendingAction
+			select {
+			case <-actionsDone:
+				return
+			case <-recvDone:
+				return
+			case pending = <-appState.pendingActions:
+			}
+
+			if pending.expiresOn.Sub(time.Now().UTC()) <= 0 {
+				// to late action already expired
+				pending.callback(nil, ErrActionTimedOut)
+				continue
+			}
+
+			appState.actionsLock.Lock()
+			err := server.Send(&proto.ActionRequest{
+				Id:     pending.id,
+				Name:   pending.name,
+				Params: pending.params,
+			})
+			if err != nil {
+				// failed to send action; add back to channel to retry on re-connect from the client
+				appState.actionsLock.Unlock()
+				appState.pendingActions <- pending
```

Comment:
food for thought: can out of order application of actions be an issue? [not a blocker]

Reply:
you have A1, A2, A3, 
A1 fails so it is sent to pendingActions channel which is buffered up to 100 items. then you will proceed with A2 and A3 and then A1 is there again from channel

Reply:
@michalpristas is correct on the ordering. But its not an issue because the actions on the client side are not blocking when it comes from reading from the stream.

So even if it goes A2, A3, A1 all 3 will be executed at the same time. Now on Agent side the `PerformAction` is blocking, even though the communication is not.

So the order of actions is still serial on Agent side:

```
PerformAction("action1") // block waiting for response
PerformAction("action2"). // this wont even be added to the channel until action1 completes, fails, or timesout
```

Reply:
Thanks for the explanation, :+1: